channels/channelmapper.go

```diff
@@ -60,79 +113,149 @@ func GetDefaultSyncFunction(scopeName, collectionName string) string {
-	numberFixBody := ConvertJSONNumbers(body)
-	numberFixMetaMap := ConvertJSONNumbers(metaMap)
+func (mapper *ChannelMapper) Function() string {
+	return mapper.fnSource
+}
+
+// This function is DEPRECATED. It's currently used in some tests that can't easily be changed.
+// Its current implementation is a kludge, and it shouldn't be used in production.
+func (mapper *ChannelMapper) SetFunction(fnSource string) error {
+	host := mapper.service.Host()
+	mapper.fnSource = fnSource
+	mapper.service = js.NewService(host, "channelMapper", wrappedFuncSource(fnSource, host))
+	return nil
+}
+
+func wrappedFuncSource(funcSource string, host js.ServiceHost) string {
+	// Use different wrapper scripts for modern vs older JS.
+	script := kSyncFnHostScriptOlder
+	if host.Engine().LanguageVersion() >= js.ES2015 {
+		script = kSyncFnHostScriptModern
+	}
+	return fmt.Sprintf(
+		script,
+		funcSource,
+		base.SyncFnErrorAdminRequired,
+		base.SyncFnErrorWrongUser,
+		base.SyncFnErrorMissingRole,
+		base.SyncFnErrorMissingChannelAccess,
+	)
+}
 
-	result1, err := mapper.Call(numberFixBody, sgbucket.JSONString(oldBodyJSON), numberFixMetaMap, userCtx)
+// Runs the sync function. Thread-safe.
+func (mapper *ChannelMapper) MapToChannelsAndAccess(body map[string]any, oldBodyJSON string, metaMap MetaMap, userCtx map[string]interface{}) (*ChannelMapperOutput, error) {
+	bodyJSON, err := json.Marshal(body)
 	if err != nil {
 		return nil, err
 	}
-	output := result1.(*ChannelMapperOutput)
-	return output, nil
+	docID, _ := body["_id"].(string)
+	revID, _ := body["_rev"].(string)
+	input := ChannelMapperInput{
+		DocID:   docID,
+		RevID:   revID,
+		Body:    string(bodyJSON),
+		OldBody: oldBodyJSON,
+		Meta:    metaMap,
+		UserCtx: userCtx,
+	}
+	return mapper.MapToChannelsAndAccess2(input)
 }
 
-// Javscript max integer value (https://www.ecma-international.org/ecma-262/5.1/#sec-8.5)
-const JavascriptMaxSafeInt = int64(1<<53 - 1)
-const JavascriptMinSafeInt = -JavascriptMaxSafeInt
-
-// ConvertJSONNumbers converts json.Number values to javascript number objects for use in the sync
-// function.  Integers that would lose precision are left as json.Number, as are floats that can't be
-// converted to float64.
-func ConvertJSONNumbers(value interface{}) interface{} {
-	switch value := value.(type) {
-	case json.Number:
-		if asInt, err := value.Int64(); err == nil {
-			if asInt > JavascriptMaxSafeInt || asInt < JavascriptMinSafeInt {
-				// Integer will lose precision when used in javascript - leave as json.Number
-				return value
-			}
-			return asInt
-		} else {
-			numErr, _ := err.(*strconv.NumError)
-			if numErr.Err == strconv.ErrRange {
-				return value
-			}
+// Runs the sync function. Thread-safe.
+func (mapper *ChannelMapper) MapToChannelsAndAccess2(input ChannelMapperInput) (*ChannelMapperOutput, error) {
```

Comment:
change the name of this function to `runSync` or `RunSync`?

Reply:
Good idea, how about `Run`?
I'll also move the old `MapToChannelsAndAccess` into `channelmapper_tests.go` since it's only used by tests.